src/Connection/DefaultConnectionPool.php

```diff
@@ -82,121 +96,27 @@ public function getStream(Request $request, CancellationToken $cancellation): Pr
-                        } elseif ($tlsState !== EncryptableSocket::TLS_STATE_ENABLED) {
-                            $socket->close();
-                            throw new SocketException('Failed to setup TLS connection, connection was in an unexpected TLS state (' . $tlsState . ')');
-                        }
-                    } catch (StreamException $exception) {
-                        $socket->close();
-                        throw new SocketException(\sprintf(
-                            "Connection to '%s' closed during TLS handshake",
-                            $authority
-                        ), 0, $exception);
-                    } catch (CancelledException $e) {
-                        $socket->close();
-
-                        // In case of a user cancellation request, throw the expected exception
-                        $cancellation->throwIfRequested();
-
-                        // Otherwise we ran into a timeout of our TimeoutCancellationToken
-                        throw new TimeoutException(\sprintf(
-                            "TLS handshake with '%s' @ '%s' timed out, took longer than " . $request->getTlsHandshakeTimeout() . ' ms',
-                            $authority,
-                            $socket->getRemoteAddress()->toString()
-                        )); // don't pass $e
-                    }
-
-                    $tlsInfo = $socket->getTlsInfo();
-                    \assert($tlsInfo !== null);
-
-                    if ($tlsInfo->getApplicationLayerProtocol() === 'h2') {
-                        $connection = new Http2Connection($socket);
-                        yield $connection->initialize();
-                    } else {
-                        if (!\array_intersect($request->getProtocolVersions(), ['1.0', '1.1'])) {
-                            $socket->close();
-                            throw new HttpException('Downgrade to HTTP/1.x forbidden, but server does not support HTTP/2');
-                        }
-
-                        $connection = new Http1Connection($socket);
-                    }
-                } else {
-                    $connection = new Http1Connection($socket);
-                }
-
-                \assert($promise instanceof Promise);
-
-                $connection->onClose(function () use ($key, $promise) {
-                    $this->connections[$key]->detach($promise);
-
-                    if (!$this->connections[$key]->count()) {
-                        unset($this->connections[$key]);
-                    }
-                });
+                return $connection->getStream($request);
+            }
 
-                return $connection;
-            });
+            $promise = new Coroutine($this->createConnection($request, $cancellation, $authority, $isHttps));
 
-            $this->connections[$key] = $this->connections[$key] ?? new \SplObjectStorage;
+            $this->connections[$key] = $this->connections[$key] ?? $connections;
```

Comment:
I think the `new \SplObjectStorage` was better here.

Reply:
`SplObjectStorage` and foreach didn't work the way I thought, so I pushed an alternative version.